### PR TITLE
Fix require (and test) performance

### PR DESCRIPTION
**Pull request by jeddy3**
Ref #3236 
Ref #2454

### CODE REVIEW

**Comment by jeddy3** [comment]
These changes take about a 3rd off the test run times on the CIs and on my local machine:

- Travis Node 8 (with coverage) 204.411s -> 166.244s
- Travis Node 6: 189.027s -> 133.654s
- Appveyor: 197.769s -> 116.166s
- Local machine (with cleaned jest cache): 248.824s -> 82.653s

These changes also reduce the `require` time of stylelint from 900ms to 600ms on my machine.

I've done a fair amount of investigation in the past two days and I've narrowed down two other possible areas of improvements:

1. Requiring `postcss-html` in `getPostcssResult.js` is responsible for nearly half of the remaining require time. Without `postcss-html` the require time is about 300ms. Which I believe is on par with ESLint. Additionally, running the tests only takes 54s on my local machine without `postcss-html`. @gucong3000 Any thoughts on what we can do about this?
2. It's difficult to pinpoint memory leaks, but there might be one in `isAutoprefixable.js`. The `*-no-vendor-prefix` rules popped up regularly when using `jest:detectleaks` and removing the `require('autoprefixer')` line from `isAutoprefixable.js` seem to quell them. Anyone got any ideas what we could do here?

@stylelint/core Can anyone see any potential problems with the changes in this PR? If people have time, it'd be great to get this PR tested against your project code bases.

**Comment by gucong3000** [comment]
> Any thoughts on what we can do about this?
#3243

**Comment by jeddy3** [comment]
> Any thoughts on what we can do about this?
> #3243

Awesome, thanks! I've merged that and rebased this PR. The change seems to have shaved off _another third_.

Travis Node 8 (with coverage: 204.411s -> 166.244s -> **77.806s** 
Travis Node 6: 189.027s -> 133.654s -> **99.038s**
Appveyor: 197.769s -> 116.166s -> **76.075s**

On my machine, tests (with cleaned Jest cache) are now down to **54s** and require time is around **344ms** (down from 900ms -> 600ms).

I'm removing the "WIP" from the title. I believe this PR is ready to review and test.

File: lib/utils/checkAgainstRule.js [hunk 1]
```diff
@@ -22,7 +23,7 @@ module.exports = function(
   if (!callback) throw new Error("checkAgainstRule requires a callback");
   if (!options.ruleName)
     throw new Error("checkAgainstRule requires a 'ruleName' option");
-  if (!rules[options.ruleName])
+  if (rules.indexOf(options.ruleName) === -1)
```
Comment (Arcanemagus) [inline lib/utils/checkAgainstRule.js]: I'd suggest `(!rules.includes(options.ruleName))` as a clearer alternative.

File: lib/requireRule.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+"use strict";
+
+const importLazy = require("import-lazy")(require);
+const rules = require("./rules");
+
+module.exports = function(ruleName) {
+  if (rules.indexOf(ruleName) !== -1) {
```
Comment (Arcanemagus) [inline lib/requireRule.js]: I'd suggest `(rules.includes(ruleName))` as a clearer alternative.

File: lib/utils/checkAgainstRule.js (hunk 1 above)
Comment (jeddy3) [reply lib/utils/checkAgainstRule.js]: For something reason, I'd gotten it into my head that `.includes` was introduced in node 8... [that's not the case](https://node.green/#ES2016-features-Array-prototype-includes).

I'll update now.

**Comment by jeddy3** [comment]
> -   Fixed: slow `require('stylelint')` time ([#3242](https://github.com/stylelint/stylelint/pull/3242)).

**Pull request merged**